File: LC/Medium/417.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        if not heights or not heights[0]: 
            return []
        
        m = len(heights)
        n = len(heights[0])
        pac_reachable = set()
        atl_reachable = set()
        
        def dfs(row, col, reachable):
            reachable.add((row, col))
            
            for (x, y) in [(1, 0), (0, 1), (-1, 0), (0, -1)]:
                new_row, new_col = row + x, col + y
                
                if new_row < 0 or new_row >= m or new_col < 0 or new_col >= n:
                    continue

                if (new_row, new_col) in reachable:
                    continue

                if heights[new_row][new_col] < heights[row][col]:
                    continue

                dfs(new_row, new_col, reachable)
        
        for i in range(m):
            dfs(i, 0, pac_reachable)
            dfs(i, n - 1, atl_reachable)
        for i in range(n):
            dfs(0, i, pac_reachable)
            dfs(m - 1, i, atl_reachable)
        
        return list(pac_reachable.intersection(atl_reachable))

File: LC/C++/Medium/417.cpp
class Solution {
public:
    int m, n;
    vector<vector<int>> heights;

    void dfs(int r, int c, set<pair<int, int>>& reachable) {
        reachable.insert({r, c});
        vector<pair<int, int>> dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

        for (auto [dx, dy] : dirs) {
            int nr = r + dx;
            int nc = c + dy;
            
            if (nr < 0 || nr >= m || nc < 0 || nc >= n) continue;
            if (reachable.count({nr, nc})) continue;
            if (heights[nr][nc] < heights[r][c]) continue;
            
            dfs(nr, nc, reachable);
        }
    }

    vector<vector<int>> pacificAtlantic(vector<vector<int>>& inputHeights) {
        heights = inputHeights;
        m = heights.size();
        n = heights.at(0).size();

        set<pair<int, int>> atlReachable;
        set<pair<int, int>> pacReachable;

        for (int i = 0; i < m; i++) {
            dfs(i, 0, pacReachable);
            dfs(i, n - 1, atlReachable);
        }

        for (int i = 0; i < n; i++) {
            dfs(0, i, pacReachable);
            dfs(m-1, i, atlReachable);
        }

        vector<vector<int>> result;
        for (auto& cell : pacReachable) {
            if (atlReachable.count(cell)) {
                result.push_back({cell.first, cell.second});
            }
        }

        return result;
    }
};